BJS2-77996

Ramil23-bit opened the pull request:
added data about work schedule, first part

Ramil23-bit committed:
added data about work schedule, first part

psvm-dev commented on src/main/java/school/faang/user_service/dto/WorkScheduleDto.java:
```diff
@@ -0,0 +1,17 @@
+    private LocalTime startLunch;
+    private LocalTime endLunch;
+    //Вот здесь вопрос, пока не понял что сделать. Надеюсь разберусь
+    private String timezone; // часовой пояс в формате IANA (например, "Europe/Moscow")
```
Отличное замечание по поводу часовых поясов. Надо разобраться, как нам нужно и потом найти, как это сделать))

psvm-dev commented on src/main/java/school/faang/user_service/exception/UserNotFoundException.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
```
500 статус используется когда произошла ошибка на стороне сервера. Если клиент запросил информацию о не существующем юзере, то ошибка на его стороне. Думаю лучше отдавать статус 404

psvm-dev commented on src/main/java/school/faang/user_service/controller/WorkScheduleController.java:
```diff
@@ -0,0 +1,26 @@
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/work_schedule")
+public class WorkScheduleController {
```
Предлагаю для контроллера тоже создать интерфейс, в котором будет описано что должен делать контроллер и чтобы этот класс его имплементировал и реализовал.

psvm-dev commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,59 @@
+    @Override
+    public WorkScheduleDto addWorkSchedule(long userId, WorkScheduleDto workScheduleDto) {
+        User user = userRepository.findById(userId).orElseThrow(() ->
+                new UserNotFoundException("User by id " + userId + " not Found"));
```
Лучше используй String.formatter("text %s text", userId). С конкатенацией тут создаётся 4 объекта стринг вместо 1.

psvm-dev commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,59 @@
+        return workScheduleDto;
+    }
+
+    public boolean isDateCheck(WorkScheduleDto workScheduleDto) {
```
Из названия метода вытекает, что он должен проверять, является ли значение датой. Но из реализации метода вижу, что метод валидирует рабочий график. Надо бы ещё подумать над неймингом.
И если это валидация, наверно надо бы её поместить в другом месте. Либо если тут оставлять, то он должен быть private, т.к. является вспомогательным методом

Keriod commented on src/main/java/school/faang/user_service/dto/WorkScheduleDto.java:
```diff
@@ -0,0 +1,17 @@
+    private LocalTime endLunch;
+    //Вот здесь вопрос, пока не понял что сделать. Надеюсь разберусь
+    private String timezone; // часовой пояс в формате IANA (например, "Europe/Moscow")
+    private Long userId;
```
Может я не до конца понял и ошибаюсь, но выглядит так, что у тебя поле userId имеет тип Long, а у WorkSchedule поле user типа User и получается маппинг разных типов без функции приведения.

FROLOV-KI commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,59 @@
+    public WorkScheduleDto addWorkSchedule(long userId, WorkScheduleDto workScheduleDto) {
+        User user = userRepository.findById(userId).orElseThrow(() ->
+                new UserNotFoundException("User by id " + userId + " not Found"));
+        if (isDateCheck(workScheduleDto)) {
```
Я бы наверное убрал этот if вообще. В методе с проверкой сделал бы void. Инвертировал бы переменные, проверяя, что данные НЕ верные. И на каждую конкретную ошибку в данных выбрасывал бы исключение. Если ни одного исключения выброшено не было, то метод отработает дальше, либо получим конкретное исключение и пользователю дадим информацию, что он не так сделал

FROLOV-KI commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,59 @@
+        LocalTime endLunch = workScheduleDto.getEndLunch();
+        LocalTime endTime = workScheduleDto.getEndTIme();
+
+        boolean startTimeAndLunch = startTime.isBefore(startLunch);
```
Я бы назвал эти переменные более конкретно, а не просто через and. 
isStartBeforLaunch, isLaunchBeforeLaunchEnd, isLaunchTimeBeforeEnd

FROLOV-KI commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,59 @@
+        if (startTimeAndLunch && startLunchAndEndLunch && endLunchAndEndTime) {
+            return true;
+        } else {
+            new DataValidationException("Неккоректные данные");
```
Ты создаешь экземпляр исключения, но не выбрасываешь его? Что значит некорректные данные? Как пользователю понять какие корректные?

FROLOV-KI commented on src/main/java/school/faang/user_service/exception/DataValidationException.java:
```diff
@@ -0,0 +1,7 @@
+package school.faang.user_service.exception;
```
Думаю вот тут лучше добавить еще один package, чтобы разделять исключения между сущностями, будет общая папка с исключениями, а в ней исключения будут разбиты по сущностям, к которым относятся, так будет проще искать нужные

psvm-dev commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,59 @@
+    private final WorkScheduleMapper workScheduleMapper;
+
+    @Override
+    public WorkScheduleDto addWorkSchedule(long userId, WorkScheduleDto workScheduleDto) {
```
метод должен быть Transactional

Ramil23-bit committed:
added update workSchedule, getByIdWorkSchedule.

FROLOV-KI commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -21,39 +24,63 @@ public class WorkScheduleServiceImpl implements WorkScheduleService {
+        WorkSchedule workSchedule = workScheduleMapper.toWorkSchedule(workScheduleCreateDto);
+        workSchedule.setUser(user);
+        checkCorrectedWorkSchedule(workScheduleCreateDto);
+        try {
```
А зачем тут этот try|catch?

Ramil23-bit committed:
fixed code.

renamed variables in method checkCorrectedWorkSchedule.
Added the package to exception

20cognosce commented on src/main/java/school/faang/user_service/controller/WorkScheduleController.java:
```diff
@@ -0,0 +1,40 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/work_schedule")
```
_ не используется в URL
/api/v1/work-schedule

20cognosce replied:
не исправлено

20cognosce commented on src/main/java/school/faang/user_service/controller/WorkScheduleController.java:
```diff
@@ -0,0 +1,40 @@
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/work_schedule")
+public class WorkScheduleController {
+    private final WorkScheduleService workScheduleService;
+    @PostMapping("/create")
```
отступ между строками

"create" не должно быть в URL - методы http POST/GET/DELETE обозначают действия

20cognosce commented on src/main/java/school/faang/user_service/controller/WorkScheduleController.java:
```diff
@@ -0,0 +1,40 @@
+    public ResponseEntity<WorkScheduleCreateDto> addWorkSchedule(@RequestBody WorkScheduleCreateDto workScheduleCreateDto) {
+        return ResponseEntity.status(HttpStatus.CREATED)
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(workScheduleService.addWorkSchedule(workScheduleCreateDto));
```
в сервис передававать сущности и примитивы, не DTO
сервис также возвращает DTO

20cognosce replied:
маппинг делать в контроллере

20cognosce commented on src/main/java/school/faang/user_service/controller/WorkScheduleController.java:
```diff
@@ -0,0 +1,40 @@
+        return workScheduleService.getById(workScheduleId);
+    }
+
+    @PutMapping("/update")
```
Patch и действие в URL не пишется

20cognosce commented on src/main/java/school/faang/user_service/dto/WorkScheduleCreateDto.java:
```diff
@@ -0,0 +1,26 @@
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+public class WorkScheduleCreateDto {
```
@Builder используй

20cognosce commented on src/main/java/school/faang/user_service/exception/users/UserNotFoundException.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
```
Этому здесь не место

20cognosce commented on src/main/java/school/faang/user_service/mapper/WorkScheduleMapper.java:
```diff
@@ -0,0 +1,15 @@
+
+@Mapper(componentModel = "spring")
+public interface WorkScheduleMapper {
+    WorkSchedule toWorkSchedule(WorkScheduleCreateDto workScheduleCreateDto);
+    WorkScheduleCreateDto toWorkScheduleDto(WorkSchedule workSchedule);
+    void update(@MappingTarget WorkSchedule workSchedule, WorkScheduleCreateDto workScheduleCreateDto);
```
ты проверил, что все поля правильно друг на друга мапятся? Для уверенного контроля может не использовать MapStruct, а руками написать методы маппинга

20cognosce replied:
либо написать в интерфейсе дефолтные методы и передать их мапстракту

20cognosce commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+public class WorkScheduleServiceImpl implements WorkScheduleService {
+    private final UserRepository userRepository;
+    private final WorkScheduleRepository workScheduleRepository;
+    private final WorkScheduleMapper workScheduleMapper;
```
в сервисе не должно быть зависимости на него

Ramil23-bit replied:
@20cognosce маппер использовать только в контроллере? а в сервисе методы должны возвращать Дто?

Ramil23-bit replied:
@20cognosce не пойму пока как мне из сущности без маппера возвращать дто

20cognosce replied:
@Ramil23-bit вопросы в дискорде или на встречах, я не читаю уведомления от гитхаба

20cognosce commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+
+    @Override
+    @Transactional
+    public WorkScheduleCreateDto addWorkSchedule(WorkScheduleCreateDto workScheduleCreateDto) {
```
Принимать сущность, а не DTO

20cognosce commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+    public WorkScheduleCreateDto addWorkSchedule(WorkScheduleCreateDto workScheduleCreateDto) {
+        long userId = userContext.getUserId();
+        User user = userRepository.findById(userId).orElseThrow(() ->
+                new UserNotFoundException("User by id " + userId + " not Found"));
```
String.format

20cognosce commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+        WorkSchedule workSchedule = workScheduleMapper.toWorkSchedule(workScheduleCreateDto);
+        workSchedule.setUser(user);
+        checkCorrectedWorkSchedule(workScheduleCreateDto);
+        try {
```
Бесполезный try catch

20cognosce commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+package school.faang.user_service.service;
+
+import jakarta.transaction.Transactional;
```
спринговую используй аннотацию

20cognosce commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+
+    @Override
+    @Transactional
+    public WorkScheduleCreateDto updateWorkScheduleDto(WorkScheduleCreateDto workScheduleCreateDto) {
+        long userId = userContext.getUserId();
+        User user = userRepository.findById(userId).orElseThrow(() ->
+                new UserNotFoundException(("User by Id not Found " + userId)));
+        WorkSchedule workSchedule = user.getWorkSchedule();
+        checkCorrectedWorkSchedule(workScheduleCreateDto);
+        workScheduleMapper.update(workSchedule, workScheduleCreateDto);
+        try {
+            workSchedule = workScheduleRepository.save(workSchedule);
+        } catch (Exception e) {
+            log.error("Error saving work schedule " + e);
+        }
+        return workScheduleMapper.toWorkScheduleDto(workSchedule);
```
аналогично

20cognosce commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Override
+    public WorkScheduleCreateDto getById(long workScheduleId) {
+        WorkSchedule workSchedule = workScheduleRepository.findById(workScheduleId)
+                .orElseThrow(() ->
+                        new WorkScheduleNotFoundException(String.format("Work schedule not Found ",
+                                workScheduleId)));
+        return workScheduleMapper.toWorkScheduleDto(workSchedule);
```
аналогично

20cognosce commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+        return workScheduleMapper.toWorkScheduleDto(workSchedule);
+    }
+
+    private void checkCorrectedWorkSchedule(WorkScheduleCreateDto workScheduleCreateDto) {
+        LocalTime startTime = workScheduleCreateDto.getStartTime();
+        LocalTime startLunch = workScheduleCreateDto.getStartLunch();
+        LocalTime endLunch = workScheduleCreateDto.getEndLunch();
+        LocalTime endTime = workScheduleCreateDto.getEndTime();
+
+        boolean isStartBeforeLaunch = startTime.isBefore(startLunch);
+        boolean isLaunchBeforeLaunchEnd = startLunch.isBefore(endLunch);
+        boolean isLaunchTimeBeforeEnd = endLunch.isBefore(endTime);
+
+        if (!(isStartBeforeLaunch && isLaunchBeforeLaunchEnd && isLaunchTimeBeforeEnd)) {
+            throw new DataValidationException("Не верно введено время работы");
+        }
```
в отдельный класс-валидатор, сделать статический метод

Ramil23-bit committed:
fixed code.

fixed the update method, getting by id

20cognosce commented on src/main/java/school/faang/user_service/controller/WorkScheduleController.java:
```diff
@@ -0,0 +1,50 @@
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public ResponseEntity<WorkScheduleCreateDto> addWorkSchedule( @RequestBody WorkSchedule workSchedule) {
+        WorkScheduleCreateDto workScheduleCreateDto = workScheduleMapper.toWorkScheduleDto(workScheduleService
+                .addWorkSchedule(workSchedule));
```
две переменные создавай

20cognosce commented on src/main/java/school/faang/user_service/controller/WorkScheduleController.java:
```diff
@@ -0,0 +1,50 @@
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public ResponseEntity<WorkScheduleCreateDto> addWorkSchedule( @RequestBody WorkSchedule workSchedule) {
```
Нет, на вход приходит DTO
ты мапишь её в POJO и передаешь в сервис, сервис отдает POJO (Entity)
тут ты обратно мапишь в DTO на ответ

20cognosce commented on src/main/java/school/faang/user_service/controller/WorkScheduleController.java:
```diff
@@ -0,0 +1,50 @@
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public ResponseEntity<WorkScheduleCreateDto> addWorkSchedule( @RequestBody WorkSchedule workSchedule) {
+        WorkScheduleCreateDto workScheduleCreateDto = workScheduleMapper.toWorkScheduleDto(workScheduleService
```
WorkScheduleCreateDto это вход
WorkScheduleDto это выход

20cognosce commented on src/main/java/school/faang/user_service/controller/WorkScheduleController.java:
```diff
@@ -0,0 +1,50 @@
+                .getById(workScheduleId));
+    }
+
+    @PutMapping
```
Patch

20cognosce commented on src/main/java/school/faang/user_service/controller/WorkScheduleController.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @PutMapping
+    public WorkScheduleUpdateDto updateWorkSchedule(@RequestBody WorkSchedule workSchedule) {
```
Аналогично

20cognosce commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Override
+    public WorkSchedule getById(long workScheduleId) {
```
Transactional(readOnly = true)

20cognosce commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,65 @@
+        WorkSchedule workScheduleDataBase = user.getWorkSchedule();
+        WorkScheduleValidation.checkCorrectedWorkSchedule(newWorkScheduleData);
+
+        workScheduleDataBase.setStartTime(newWorkScheduleData.getStartTime());
+        workScheduleDataBase.setStartLunch(newWorkScheduleData.getStartLunch());
+        workScheduleDataBase.setEndLunch(newWorkScheduleData.getEndLunch());
+        workScheduleDataBase.setEndTime(newWorkScheduleData.getEndTime());
+        workScheduleDataBase.setTimezone(newWorkScheduleData.getTimezone());
```
если я захотел обновить только время обеда, остальные поля передал null, ты в БД их сотрешь получается?

Ramil23-bit committed:
fixed code.

added new Dto,fixed mapper, service and controller

20cognosce commented on src/main/java/school/faang/user_service/controller/WorkScheduleController.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @PatchMapping
+    public WorkScheduleUpdateDto updateWorkSchedule(@RequestBody WorkScheduleUpdateDto workScheduleUpdateDto) {
```
почему ты здесь возвращаешь WorkScheduleUpdateDto...
надо возвращать WorkScheduleDto

20cognosce commented on src/main/java/school/faang/user_service/service/WorkScheduleServiceImpl.java:
```diff
@@ -0,0 +1,73 @@
+        WorkSchedule workScheduleDataBase = user.getWorkSchedule();
+        WorkScheduleValidation.checkCorrectedWorkSchedule(newWorkScheduleData);
+
+        executeIfNotNull(newWorkScheduleData.getStartTime(),
+                () -> workScheduleDataBase.setStartTime(newWorkScheduleData.getStartTime()));
```
можно было в 1 строчку
